http/peer-id-auth.md

```diff
@@ -0,0 +1,101 @@
+# Peer ID Authentication over HTTP
+
+| Lifecycle Stage | Maturity      | Status | Latest Revision |
+| --------------- | ------------- | ------ | --------------- |
+| 1A              | Working Draft | Active | r0, 2023-01-23  |
+
+Authors: [@MarcoPolo]
+
+Interest Group: Same as [HTTP](README.md)
+
+## Introduction
+
+This spec defines one way of authenticating Peer IDs over HTTP using a
+challenge-response scheme.
+
+## Mutual Client and Server Peer ID Authentication
+
+1. The server initiates the authentication by responding to a request that must
+   be authenticated with the response header `WWW-Authenticate: Libp2p-Challenge
+   challenge="<base64-encoded-challenge>, Libp2p-Challenge-Server-Only"`. The
+   challenge MUST be randomly generated from server for sole purpose of
+   authenticating the client. The server SHOULD store the challenge temporarily
```

Comment:
This breaks the stateless nature of HTTP. Subsequent requests might end up at a different backend server.

Reply:
This is what webauthn does. I copied it from there.

An alternative is to use some encrypted value from the server, but this opens you up to replay attacks. They can be mitigated by including a validity time window. I went with the less risky approach that is widely deployed in webauthn.

Reply:
> This breaks the stateless nature of HTTP. Subsequent requests might end up at a different backend server.

I don't think this is true. HTTP being stateless means that each message is entirely self-contained and doesn't depend on _who_ sends the message.

For example, FTP is not stateless because it contains commands like `LIST` which do different things depending on which directory the client is in. That is a stateful protocol because the server has to keep state _per client_ to actually fulfill the request.

Depending on server-side state for an application protocol does not break HTTP's stateless aspect.

If you load balance an HTTP interface, it is fairly well understood that you need to share application state between those to "hide" the load balancing from the client. For example, if a client first creates a resource via `POST` and then tries to `GET` it, your application needs to ensure that this works despite the load balancing.

Reply:
It depends on how the application is structured. There are cases where you won't read your writes (GET after a POST will fail temporarily until some backend is consistent).

I think auth is an exception, and it's okay to store some state temporarily to avoid replay attacks. This is the same thing WebAuthn does: 
> As a cryptographic protocol, Web Authentication is dependent upon randomized challenges to avoid replay attacks. Therefore, the values of both [PublicKeyCredentialCreationOptions](https://w3c.github.io/webauthn/#dictdef-publickeycredentialcreationoptions).[challenge](https://w3c.github.io/webauthn/#dom-publickeycredentialcreationoptions-challenge) and [PublicKeyCredentialRequestOptions](https://w3c.github.io/webauthn/#dictdef-publickeycredentialrequestoptions).[challenge](https://w3c.github.io/webauthn/#dom-publickeycredentialrequestoptions-challenge) MUST be randomly generated by [Relying Parties](https://w3c.github.io/webauthn/#relying-party) in an environment they trust (e.g., on the server-side), and the returned [challenge](https://w3c.github.io/webauthn/#dom-collectedclientdata-challenge) value in the client’s response MUST match what was generated. This SHOULD be done in a fashion that does not rely upon a client’s behavior, e.g., the [Relying Party](https://w3c.github.io/webauthn/#relying-party) SHOULD store the challenge temporarily until the operation is complete. Tolerating a mismatch will compromise the security of the protocol.

https://w3c.github.io/webauthn/#sctn-cryptographic-challenges

Reply:
Thinking about this some more, applications should be able to decide where they want to fall on the replay protection spectrum. If they want no possibility of replay in exchange for some auth state, they could randomly generate the token. If they are okay with a small window of replay (maybe that's fine for the application), they could sign a timestamp and a nonce and use that to statelessly authenticate the client.

With that in mind, I'll change the wording to allow both, and applications/implementations can decide what they want to do.